stl/inc/xspan

```diff
@@ -0,0 +1,238 @@
+    // ~span() = default;
+
+    // [span.sub], span subviews
+    template <size_t _Count>
+    constexpr span<element_type, _Count> first() const noexcept {
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Count <= _Extent, "Count out of range in span::first()");
+        } else {
+#if _CONTAINER_DEBUG_LEVEL > 0
+            _STL_VERIFY(_Count <= size(), "Count out of range in span::first()");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        }
+        return {data(), _Count};
+    }
+
+    constexpr span<element_type, dynamic_extent> first(index_type _Count) const noexcept {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Count <= size(), "Count out of range in span::first(Count)");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return {data(), _Count};
+    }
+
+    template <size_t _Count>
+    constexpr span<element_type, _Count> last() const noexcept {
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Count <= _Extent, "Count out of range in span::last()");
+        } else {
+#if _CONTAINER_DEBUG_LEVEL > 0
+            _STL_VERIFY(_Count <= size(), "Count out of range in span::last()");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        }
+        return {data() + size() - _Count, _Count};
+    }
+
+    constexpr span<element_type, dynamic_extent> last(index_type _Count) const noexcept {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Count <= size(), "Count out of range in span::first(Count)");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return {data() + size() - _Count, _Count};
+    }
+
+    template <size_t _Offset, size_t _Count = dynamic_extent>
+    constexpr span<element_type, dynamic_extent> subspan() const noexcept {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Offset <= size(), "Offset out of range in span::subspan()");
+        _STL_VERIFY(_Count == dynamic_extent || _Offset + _Count <= size(), "Count out of range in span::subspan()");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return {data() + _Offset, _Count == dynamic_extent ? size() - _Offset : _Count};
+    }
+
+    constexpr span<element_type, dynamic_extent> subspan(index_type _Offset, index_type _Count = dynamic_extent) const
+        noexcept {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Offset <= size(), "Offset out of range in span::subspan(offset, count)");
+        _STL_VERIFY(_Count == dynamic_extent || _Offset + _Count <= size(),
+            "Count out of range in span::subspan(offset, count)");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return {data() + _Offset, _Count == dynamic_extent ? size() - _Offset : _Count};
+    }
+
+    // [span.obs], span observers
+    constexpr index_type size() const noexcept {
+        return _Mypair._Get_first().size();
+    }
+
+    constexpr index_type size_bytes() const noexcept {
+        return size() * static_cast<index_type>(sizeof(element_type));
+    }
+
+    _NODISCARD constexpr bool empty() const noexcept {

```

Comment:
This entire file needs to be audited for `_NODISCARD`. We massively strengthen nodiscard compared to what's depicted in the Standard. We have several criteria:

* Pure observers (functions that have no effect other than to return a value)
* Things which would leak resources if the return value were discarded (e.g. allocation functions)
* Things where discarding the return value is extremely likely to lead to incorrect code (rare; mostly remove_if/unique)

Here, things like `size()` and `operator[]` are pure observers that should be marked, among many others.

Reply:
How about constructors as that is coming too?

Reply:
> How about constructors as that is coming too?

Woops. That's so recent a change I forgot to exclude constructors from my comment. I think only Stephan's third bullet will apply when considering application of `[[nodiscard]]` to constructors, but admittedly we haven't yet given it much thought. #63 is the "think hard about precise criteria for marking constructors `_NODISCARD` and then apply those criteria to the STL" issue.

Dropping `span`s on the floor is at worse a loss of performance, never a correctness bug, so we  should not tag its constructors.

Reply:
I assume this discussion as resolved with slapping `_NODISCARD` everywhere except constructors